### PR TITLE
Zarr Reading

### PR DESCRIPTION
**Pull request by aidang3019**

### CODE REVIEW

**Comment by aidang3019** [comment]
* **#198** <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/198?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 5  dependent PRs ([#199](https://github.com/GaTech-RL2/EgoVerse/pull/199) <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/199?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#200](https://github.com/GaTech-RL2/EgoVerse/pull/200) <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/200?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#208](https://github.com/GaTech-RL2/EgoVerse/pull/208) <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/208?utm_source=stack-comment-icon" target="_blank">Graphite</a> and 2 others)
* **#197** <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/197?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/197?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: egomimic/rldb/zarr/action_chunk_transforms.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    def transform_joints(self, actions: np.ndarray, chunk_length: int) -> np.ndarray:
+        raise NotImplementedError
+
+
+# ---------------------------------------------------------------------------
+# Concrete transform classes
+# ---------------------------------------------------------------------------
+
+class SingleArmEulerTransform(BaseActionChunkTransform):
+    """Single-arm 6D (xyz + euler ypr) transform."""
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/action_chunk_transforms.py]: clearer docstring (input vs output_

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+    return EMBODIMENT_ID_TO_KEY.get(index, None)
+
+
+def get_embodiment_id(embodiment_name):
+    embodiment_name = embodiment_name.upper()
+    return EMBODIMENT[embodiment_name].value
+
+
+
+class EpisodeResolver:
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py:78]: Have abstract episode resolver which just has the static class funcs, then inherit LocalEpisodeResolver and S3EpisodeResolver which implement their own resolve functions

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+    @classmethod
+    def _load_zarr_datasets(cls, *, search_path: Path, valid_hashes: set[str], action_horizon = 100):
+        
+        """
+        Loads multiple Zarr datasets from the specified folder path, filtering only those whose hashes
+        are present in the valid_hashes set.
+
+        Args:
+            folder_path (Path): The root folder containing candidate dataset subfolders.
+            valid_hashes (set[str]): Set of valid dataset hashes (names) to load.
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: rename clearer

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+                ds_obj = ZarrDataset(p, action_horizon=action_horizon)
+                datasets[name] = ds_obj
+            except Exception as e:
+                logger.error(f"Failed to load dataset at {p}: {e}")
+                skipped.append(p.name)
+            
+        return datasets, skipped
+
+    @staticmethod
+    def _get_filtered_paths(filters):
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: This is s3 resolver specific

File: egomimic/rldb/zarr/action_chunk_transforms.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
+"""
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/action_chunk_transforms.py:1]: run ruff on everything

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 6]
```diff
@@ -0,0 +1,741 @@
...
+
+
+class MultiDataset(torch.utils.data.Dataset):
+    """
+    Self wrapping MultiDataset, can wrap zarr or multi dataset. 
+
+    """
+    def __init__(self, 
+        datasets,
+        embodiment,
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: Multi should not have embodiment.  See where this breaks the training code.  In ZarrDataset/episode it should have robot_name not embodiment.  And each batch should probably also return the EmbodimentName id -> robot_name id in the batch itself.

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 7]
```diff
@@ -0,0 +1,741 @@
...
+
+
+
+class EpisodeResolver:
+    """
+    Filters SQL table for zarr episode paths/ downloads from S3.
+    resolve returns processed_path. 
+    """
+    def __init__(
+        self,
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py:84]: For all function args, show typing for arguments.

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 8]
```diff
@@ -0,0 +1,741 @@
...
+                n_keep = max(1, n_keep)
+            chosen = set(all_names[:n_keep])
+        else:
+            raise ValueError(f"Unknown mode: {mode}")
+
+        datasets = {rid: ds for rid, ds in datasets.items() if rid in chosen}
+        assert datasets, "No datasets left after applying mode split."
+
+        self.key_map = (
+            {repo_id: key_map for repo_id in datasets} if key_map else None
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: rm repo id

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 9]
```diff
@@ -0,0 +1,741 @@
...
+        dataset_name, local_idx = self.index_map[idx]
+        data = self.datasets[dataset_name][local_idx]
+
+        if self.key_map and dataset_name in self.key_map:
+            key_map = self.key_map[dataset_name]
+            data = {key_map.get(k, k): v for k, v in data.items()}
+
+        return data
+
+    def _merge_datasets(self):
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: delete this

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 10]
```diff
@@ -0,0 +1,741 @@
...
+                sub_dataset = sub_dataset.rename_columns(key_map)
+
+            dataset_list.append(sub_dataset)
+
+        merged_dataset = concatenate_datasets(dataset_list)
+
+        return merged_dataset
+    
+    @classmethod
+    def _from_resolver(cls, resolver: EpisodeResolver, embodiment, action_horizon = 100, **kwargs):
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: embodiment

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 11]
```diff
@@ -0,0 +1,741 @@
...
+        filters = kwargs.pop("filters", {}) or {}
+
+        resolved = resolver.resolve(
+            embodiment=embodiment,
+            sync_from_s3=sync_from_s3,
+            action_horizon=action_horizon,
+            filters=filters,
+        )
+
+        # resolver may return datasets or (datasets, skipped)
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: resolver should either return datasets or tuple but not either

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 12]
```diff
@@ -0,0 +1,741 @@
...
+
+
+class ZarrDataset(torch.utils.data.Dataset):
+    """
+    Base Zarr Dataset object, Just intializes as pass through to read from zarr episode
+    """
+
+    def __init__(
+        self,
+        Episode_path: str,
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: all paths should be pathlib objects not strings

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 13]
```diff
@@ -0,0 +1,741 @@
...
+        Detect which keys contain JPEG-encoded image data from metadata.
+
+        Returns:
+            Set of keys containing JPEG data
+        """
+        features = self.metadata.get("features", {})
+        return {key for key, info in features.items() if info.get("dtype") == "jpeg"}
+
+    def __len__(self) -> int:
+        if hasattr(self, 'total_frames'):
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: prob get rid of the lazy stuff if not actually functional rn

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 14]
```diff
@@ -0,0 +1,741 @@
...
+                end_idx = min(idx + self.action_horizon, self.total_frames)
+                keys_dict[k] = (idx, end_idx)
+            else:
+                # Load single frame
+                keys_dict[k] = (idx, None)
+
+        data = self.episode_reader.read(keys_dict)
+
+        # Pad action sequences to fixed length if needed
+        if self.action_horizon is not None:
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: make padding a helper func

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 15]
```diff
@@ -0,0 +1,741 @@
...
+                        # Pad by repeating the last frame
+                        pad_len = self.action_horizon - seq_len
+                        last_frame = data[k][-1:]  # Keep dims: (1, action_dim)
+                        padding = np.repeat(last_frame, pad_len, axis=0)
+                        data[k] = np.concatenate([data[k], padding], axis=0)
+
+        # Apply embodiment-specific interpolation to action chunks
+        if self.action_horizon is not None and self.chunk_length is not None:
+            for k in self.action_keys:
+                if k in data and isinstance(data[k], np.ndarray):
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: np check redundant

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 16]
```diff
@@ -0,0 +1,741 @@
...
+                    data[k] = self.action_transform.transform(data[k], self.chunk_length, key=k)
+
+        # Decode JPEG-encoded image data and normalize to [0, 1]
+        import simplejpeg
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: imports all at top

File: egomimic/rldb/zarr/action_chunk_transforms.py [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+    def transform_joints(self, actions: np.ndarray, chunk_length: int) -> np.ndarray:
+        return _interpolate_linear(actions, chunk_length)
+
+    def transform(self, actions: np.ndarray, chunk_length: int, key: str) -> np.ndarray:
+        if "cartesian" in key:
+            return self.transform_cartesian(actions, chunk_length)
+        return self.transform_joints(actions, chunk_length)
+
+
+class SingleArmEulerGripTransform(BaseActionChunkTransform):
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/action_chunk_transforms.py]: Transforms fixes  
\- zarr dataset should accept a list of transforms  
\- each transform's transform function should take just a keydict, then reference the classes init variables to see how to index into that keydict.  
\- Then zarrdataset itself shouldn't have imagekeys or actionkeys passed in

File: egomimic/rldb/zarr/zarr_dataset_multi.py [hunk 18]
```diff
@@ -0,0 +1,741 @@
...
+            result[key] = data
+        return result
+    def _collect_keys(self) -> list[str]:
+        """
+        Collect all array keys from the store.
+        Returns:
+            List of array keys (flat structure with dot-separated names)
+        """
+        keys = []
+        for name in self._store.array_keys():
```
Comment (SimarKareer) [inline egomimic/rldb/zarr/zarr_dataset_multi.py]: edit to use metadata instead of store.array_keys()

**Comment by SimarKareer** [comment]
### Merge activity

* **Feb 24, 11:20 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/197).
* **Feb 24, 11:21 PM UTC**: @SimarKareer merged this pull request with [Graphite](https://app.graphite.com/github/pr/GaTech-RL2/EgoVerse/197).
